362.py

```diff
@@ -0,0 +1,45 @@
+class HitCounter:
+
+    def __init__(self):
+        """
+        Initialize your data structure here.
+        """
+        self.hits = collections.deque([])
+        self.total = 0
+
+
+    def hit(self, timestamp: int) -> None:
+        """
+        Record a hit.
+        @param timestamp - The current timestamp (in seconds granularity).
+        """
+        if len(self.hits) == 0 or self.hits[-1][0] != timestamp:
+            self.hits.append([timestamp, 1])
+        else:
+            self.hits[-1][1] += 1
+
+        self.total += 1
+
+
+    def getHits(self, timestamp: int) -> int:
+        """
+        Return the number of hits in the past 5 minutes.
+        @param timestamp - The current timestamp (in seconds granularity).
+        """
+
+        while self.hits:
```

Comment:
Good:
- You aggregated the count of each `timestamp`.

Not so good:
- You are using a linear search approach to get back to 300 seconds ago. Considering that the elements in the `deque` is already sorted, binary search should be used to locate the indexes of `timestamp - 300 + 1` and `timestamp` in the queue.
- You are discarding the old timestamps from `deque`. It is okay in this problem, but what if the requirement for `getHits()` becomes: given any timestamp (which could happen to be something that you have already discarded), return the total number of hits between `[timestamp - 300 + 1, timestamp]`?

Reply:
For not so good parts:
1. This makes sense!
2. We design based on the given requirements. If we have to support `getHits()` queries at any time, we have to trade more storage for that. We can optimize how we gonna use storage, for example, put `[300s, timestamp]` to memory and the rest to disk.

Reply:
Do you have some other good idea on point 2?